### PR TITLE
Retry `S3BlobContainer#getRegister` on all exceptions

### PR DESCRIPTION
**Pull request by DaveCTurner**
S3 register reads are subject to the regular client retry policy, but in
practice we see failures of these reads sometimes for errors that are
transient but for which the SDK does not retry. This commit adds another
layer of retries to these reads.

Relates ES-9721

### CODE REVIEW

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 1]
```diff
@@ -910,21 +911,43 @@ public void compareAndExchangeRegister(
...
+                    }
                 }
+                if (retryDelayIterator.hasNext()) {
+                    try {
+                        // noinspection BusyWait
+                        Thread.sleep(retryDelayIterator.next().millis());
+                        continue;
+                    } catch (InterruptedException interruptedException) {
+                        Thread.currentThread().interrupt();
+                        finalException.addSuppressed(interruptedException);
```
Comment (nicktindall) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:946]: Should we break out of the loop if we're interrupted? I'm guessing something else in the loop will throw an InterruptedException if we continue, but would it be more polite to bail early?

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 2]
```diff
@@ -910,21 +911,43 @@ public void compareAndExchangeRegister(
...
+                ) {
+                    return OptionalBytesReference.of(getRegisterUsingConsistentRead(stream, keyPath, key));
+                } catch (Exception attemptException) {
+                    logger.trace(() -> Strings.format("[%s]: getRegister failed", key), attemptException);
+                    if (attemptException instanceof AmazonS3Exception amazonS3Exception && amazonS3Exception.getStatusCode() == 404) {
+                        return OptionalBytesReference.EMPTY;
+                    } else if (finalException == null) {
+                        finalException = attemptException;
+                    } else if (finalException != attemptException) {
+                        finalException.addSuppressed(attemptException);
```
Comment (nicktindall) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:936]: How is it possible that `finalException == attemptException`? does something we call throw a singleton exception? I'm not sure I follow this condition.

**Review by nicktindall — APPROVED** [review]
LGTM with minor comments

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java (hunk 1 above)
Comment (DaveCTurner) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:946]: Yeah we do break out here, we fall through to the `throw finalException` line here. I added a comment.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java (hunk 2 above)
Comment (DaveCTurner) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:936]: Who knows what the SDK does or might do in future, but calling `addSuppressed` with the same exception throws an `IllegalArgumentException` and we don't want that here.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java (hunk 1 above)
Comment (nicktindall) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:946]: Ah yes

**Pull request merged**